bigml/topicmodel.py

```diff
@@ -376,6 +378,57 @@ def infer(self, list_of_indices):
         return [(sample_counts[k] + self.alpha) / normalizer
                 for k in range(self.ntopics)]
 
+    def batch_predict(self, input_data_list, outputs=None, **kwargs):
+        """Creates a batch prediction for a list of inputs using the local
+        supervised model. Allows to define some output settings to
+        decide the fields to be added to the input_data (prediction,
+        probability, etc.) and the name that we want to assign to these new
+        fields. The outputs argument accepts a dictionary with keys
+        "output_fields", to contain a list of the prediction properties to add
+        (["prediction", "probability"] by default) and "output_headers", to
+        contain a list of the headers to be used when adding them (identical
+        to "output_fields" list, by default).
+
+        :param input_data_list: List of input data to be predicted
+        :type input_data_list: list or Panda's dataframe
+        :param dict outputs: properties that define the headers and fields to
+                             be added to the input data
+        :return: the list of input data plus the predicted values
+        :rtype: list or Panda's dataframe depending on the input type in
+                input_data_list
+
+        """
+        if outputs is None:
+            outputs = {}
+        if outputs.get(OUT_NEW_FIELDS) is None:
+            new_fields = [topic["name"] for topic in self.topics]
+        if outputs.get(OUT_NEW_HEADERS):
+            new_headers = outputs.get(OUT_NEW_HEADERS, new_fields)
+            if len(new_fields) > len(new_headers):
+                new_headers.expand(new_fields[len(new_headers):])
+            else:
+                new_headers[0: len(new_fields)]
+        else:
+            new_headers = new_fields
+        dataframe = False
+        if pandas_ready and isinstance(input_data_list, DataFrame):
+            dataframe = True
+            inner_data_list = input_data_list.to_dict('records')
+        else:
+            inner_data_list = input_data_list[:]
+        for index, input_data in enumerate(inner_data_list):
+            prediction = self.distribution(input_data, **kwargs)
+            prediction_dict = {}
+            for topic_distribution in prediction:
+                prediction_dict.update({topic_distribution["name"]:
+                    topic_distribution["probability"]})
+            for ikey, key in enumerate(new_fields):
+                inner_data_list[index][new_headers[ikey]] = prediction_dict[
+                    key]
+        if pandas_ready and dataframe:
+            return DataFrame.from_dict(inner_data_list)
+        return inner_data_list
```

Comment:
this logic of checking wether the inputs are a dataframe and the applying a "prediction function" over it is repeated over and over again: would it be possible to share it, perhaps on a mixin class with a method for the specific part?